In [1]:
pip install imbalanced-learn

     |████████████████████████████████| 167 kB 4.1 MB/s eta 0:00:01
Note: you may need to restart the kernel to use updated packages.


In [1]:
import pandas as pd
url="https://bd29ee0e-54ab-4daa-9671-d153865d1620.usrfiles.com/ugd/bd29ee_96e60c86a00b486ea5130e16275c1674.csv"
df=pd.read_csv(url)
df.head()

X0   X1   X2   X3   X4   X5   X6   X7   X8   X9  ...  X170  X171  \
0  X21.V1.791  135  190  229  223  192  125   55   -9  -33  ...   -17   -15   
1  X15.V1.924  386  382  356  331  320  315  307  272  244  ...   164   150   
2     X8.V1.1  -32  -39  -47  -37  -32  -36  -57  -73  -85  ...    57    64   
3   X16.V1.60 -105 -101  -96  -92  -89  -95 -102 -100  -87  ...   -82   -81   
4   X20.V1.54   -9  -65  -98 -102  -78  -48  -16    0  -21  ...     4     2   

   X172  X173  X174  X175  X176  X177  X178  y  
0   -31   -77  -103  -127  -116   -83   -51  4  
1   146   152   157   156   154   143   129  1  
2    48    19   -12   -30   -35   -35   -36  5  
3   -80   -77   -85   -77   -72   -69   -65  5  
4   -12   -32   -41   -65   -83   -89   -73  5  

[5 rows x 180 columns]

In [2]:
df.shape

(11500, 180)

In [3]:
X=df.drop(["X0","y"],axis=1)
X.head()

X1   X2   X3   X4   X5   X6   X7   X8   X9  X10  ...  X169  X170  X171  \
0  135  190  229  223  192  125   55   -9  -33  -38  ...     8   -17   -15   
1  386  382  356  331  320  315  307  272  244  232  ...   168   164   150   
2  -32  -39  -47  -37  -32  -36  -57  -73  -85  -94  ...    29    57    64   
3 -105 -101  -96  -92  -89  -95 -102 -100  -87  -79  ...   -80   -82   -81   
4   -9  -65  -98 -102  -78  -48  -16    0  -21  -59  ...    10     4     2   

   X172  X173  X174  X175  X176  X177  X178  
0   -31   -77  -103  -127  -116   -83   -51  
1   146   152   157   156   154   143   129  
2    48    19   -12   -30   -35   -35   -36  
3   -80   -77   -85   -77   -72   -69   -65  
4   -12   -32   -41   -65   -83   -89   -73  

[5 rows x 178 columns]

In [4]:
y=df[["y"]].rename(columns={"y":"Class"})
y.head()

Class
0      4
1      1
2      5
3      5
4      5

In [5]:
import numpy as np
np.unique(y)

array([1, 2, 3, 4, 5])

In [6]:
y=y.replace([2,3,4,5],0)
np.unique(y)

array([0, 1])

In [7]:
y.groupby("Class")["Class"].count()

Class
0    9200
1    2300
Name: Class, dtype: int64

In [8]:
import numpy as np
from numpy import set_printoptions
from sklearn.feature_selection import SelectKBest
from sklearn.feature_selection import f_classif

# feature extraction. We are going to output the selection scores for all features and select the features with the highest scores.
test = SelectKBest(score_func=f_classif, k="all")
fit = test.fit(X, np.asarray(y).reshape(y.shape[0],))
features = fit.transform(X)
# summarize scores
set_printoptions(precision=3)

scores=fit.scores_
scores=pd.DataFrame(scores)
scores=scores.rename(columns={0:"scores"})
scores=scores.sort_values(by=["scores"],ascending=False)
scores

scores
10   36.886894
9    32.148024
11   30.713085
43   28.903209
157  28.082714
..         ...
121   0.008356
127   0.007285
57    0.003150
118   0.000104
165   0.000075

[178 rows x 1 columns]

In [9]:
X_scores=scores[scores["scores"]>8]
len(X_scores)

53

In [10]:
X_prime=X.iloc[:,X_scores.index]
X_prime.head()

X11  X10  X12  X44  X158  X159  X160  X43   X9   X8  ...  X124  X150  X25  \
0  -10  -38   35    9   -53   -79   -72    6  -33   -9  ...    59   -44   11   
1  237  232  258  297    63    69    89  303  244  272  ...  -224   280  118   
2  -99  -94  -94  -65    80    66    23  -53  -85  -73  ...   -80   -62  -78   
3  -72  -79  -68  -54   -38   -22   -29  -58  -87 -100  ...   -84   -64  -94   
4  -90  -59 -103  -42    11     3    -1  -14  -21    0  ...    35    57   78   

   X35  X70  X14  X82  X149  X152  X155  
0 -130   12  113    8   -65   -57  -130  
1  326 -432    2  453   307   159    43  
2    7   25 -104  -64   -63     1    37  
3  -69  -69  -80  -63   -58   -80   -83  
4  -19   13  -43    0    29    45   -23  

[5 rows x 53 columns]

In [11]:
from sklearn.preprocessing import StandardScaler
scaler=StandardScaler()
X_prime=pd.DataFrame(scaler.fit_transform(X_prime),columns=X_prime.columns)
X_prime.head()

X11       X10       X12       X44      X158      X159      X160  \
0 -0.026003 -0.198415  0.254645  0.082069 -0.288385 -0.438941 -0.394005   
1  1.513283  1.484569  1.638242  1.880314  0.403959  0.434674  0.549548   
2 -0.580644 -0.547478 -0.545732 -0.379980  0.505424  0.416966  0.162750   
3 -0.412382 -0.453979 -0.384415 -0.311297 -0.198857 -0.102481 -0.142000   
4 -0.524557 -0.329314 -0.601572 -0.236370  0.093598  0.045089  0.022096   

        X43        X9        X8  ...      X124      X150       X25       X35  \
0  0.067948 -0.163196 -0.014267  ...  0.415824 -0.220157  0.103935 -0.777606   
1  1.953854  1.546403  1.719102  ... -1.277240  1.706706  0.753569  2.058095   
2 -0.306693 -0.484131 -0.409056  ... -0.415752 -0.327205 -0.436416  0.074348   
3 -0.338442 -0.496475 -0.575607  ... -0.439682 -0.339099 -0.533557 -0.398269   
4 -0.059049 -0.089134  0.041250  ...  0.272242  0.380501  0.510715 -0.087337   

        X70       X14       X82      X149      X152      X155  
0  0.109490  0.709783  0.107112 -0.345679 -0.306837 -0.735615  
1 -2.612161  0.053720  2.708751  1.880217  1.001824  0.293312  
2  0.189178 -0.572791 -0.313828 -0.333712  0.044563  0.257627  
3 -0.387027 -0.430940 -0.307981 -0.303794 -0.446185 -0.456080  
4  0.115620 -0.212252  0.060341  0.216778  0.311142 -0.099226  

[5 rows x 53 columns]

In [12]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_prime,y,test_size=0.2,random_state=42)
print(X_train.shape)
print(X_test.shape)

(9200, 53)
(2300, 53)


In [13]:
from imblearn.ensemble import BalancedRandomForestClassifier
BForest=BalancedRandomForestClassifier(n_estimators=170,criterion="entropy",min_samples_split=3,min_samples_leaf=2,
                                       sampling_strategy="not majority",n_jobs=-1,
                                       random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
BForest

BalancedRandomForestClassifier(criterion='entropy', min_samples_leaf=2,
                               min_samples_split=3, n_estimators=170, n_jobs=-1,
                               random_state=0,
                               sampling_strategy='not majority')

In [14]:
from sklearn.metrics import accuracy_score
train_pred=BForest.predict(X_train)
test_pred=BForest.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9997826086956522
Accuracy on test data:  0.9717391304347827


In [15]:
from sklearn.metrics import f1_score
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  0.9997826086956522
F1 score on test data:  0.9717391304347827


In [16]:
from sklearn.metrics import precision_score
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  0.9997826086956522
Precision score on test data:  0.9717391304347827


In [17]:
from sklearn.metrics import recall_score
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  0.9997826086956522
Recall score on test data:  0.9717391304347827


In [18]:
from sklearn.metrics import confusion_matrix
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7365    0]
 [   2 1833]]
Confusion matrix on test data: 
 [[1816   19]
 [  46  419]]


In [52]:
from imblearn.ensemble import BalancedBaggingClassifier
from sklearn import svm
BBagging=BalancedBaggingClassifier(base_estimator=svm.SVC(C=3,gamma="auto"),n_estimators=15,sampling_strategy="not majority",
                                   random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
BBagging

BalancedBaggingClassifier(base_estimator=SVC(C=3, gamma='auto'),
                          n_estimators=15, random_state=0,
                          sampling_strategy='not majority')

In [53]:
from sklearn.metrics import accuracy_score
train_pred=BBagging.predict(X_train)
test_pred=BBagging.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9827173913043479
Accuracy on test data:  0.9734782608695652


In [54]:
from sklearn.metrics import f1_score
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  0.9827173913043479
F1 score on test data:  0.9734782608695652


In [55]:
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  0.9827173913043479
Precision score on test data:  0.9734782608695652


In [56]:
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  0.9827173913043479
Recall score on test data:  0.9734782608695652


In [57]:
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7342   23]
 [ 136 1699]]
Confusion matrix on test data: 
 [[1822   13]
 [  48  417]]


In [117]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_prime,y,test_size=0.2,random_state=42)
print(X_train.shape)
print(X_test.shape)

(9200, 53)
(2300, 53)


In [118]:
train=pd.concat([X_train,y_train],axis=1)
train0=train[train["Class"]==0]
train1=train[train["Class"]==1]
train1=pd.concat([train1]*2,axis=0)
train=pd.concat([train1,train0],axis=0)
X_train=train.drop(["Class"],axis=1)
y_train=train[["Class"]]
print(X_train.shape)

(11035, 53)


In [60]:
from sklearn.ensemble import RandomForestClassifier
Forest=RandomForestClassifier(random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Forest

RandomForestClassifier(random_state=0)

In [61]:
from sklearn.metrics import accuracy_score
train_pred=Forest.predict(X_train)
test_pred=Forest.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  1.0
Accuracy on test data:  0.97


In [62]:
Forest=RandomForestClassifier(n_estimators=200,criterion="entropy",min_samples_split=4,n_jobs=-1,
                              random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Forest

RandomForestClassifier(criterion='entropy', min_samples_split=4,
                       n_estimators=200, n_jobs=-1, random_state=0)

In [63]:
train_pred=Forest.predict(X_train)
test_pred=Forest.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  1.0
Accuracy on test data:  0.9739130434782609


In [64]:
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  1.0
F1 score on test data:  0.9739130434782609


In [65]:
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  1.0
Precision score on test data:  0.9739130434782609


In [66]:
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  1.0
Recall score on test data:  0.9739130434782609


In [67]:
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7365    0]
 [   0 3670]]
Confusion matrix on test data: 
 [[1810   25]
 [  35  430]]


In [129]:
from sklearn import svm
SVM=svm.SVC(random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
SVM

SVC(random_state=0)

In [130]:
train_pred=SVM.predict(X_train)
test_pred=SVM.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9645672859084731
Accuracy on test data:  0.971304347826087


In [149]:
SVM=svm.SVC(C=6,gamma="auto",
            random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
SVM

SVC(C=6, gamma='auto', random_state=0)

In [150]:
train_pred=SVM.predict(X_train)
test_pred=SVM.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9872224739465337
Accuracy on test data:  0.9791304347826087


In [151]:
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  0.9872224739465337
F1 score on test data:  0.9791304347826087


In [152]:
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  0.9872224739465337
Precision score on test data:  0.9791304347826087


In [153]:
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  0.9872224739465337
Recall score on test data:  0.9791304347826087


In [154]:
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7346   19]
 [ 122 3548]]
Confusion matrix on test data: 
 [[1819   16]
 [  32  433]]


In [68]:
from sklearn.ensemble import GradientBoostingClassifier
Gradient=GradientBoostingClassifier(random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Gradient

GradientBoostingClassifier(random_state=0)

In [69]:
train_pred=Gradient.predict(X_train)
test_pred=Gradient.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9710013593112823
Accuracy on test data:  0.96


In [70]:
Gradient=GradientBoostingClassifier(n_estimators=210,learning_rate=0.15,max_depth=7,max_features="log2",
                                    random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Gradient

GradientBoostingClassifier(learning_rate=0.15, max_depth=7, max_features='log2',
                           n_estimators=210, random_state=0)

In [71]:
train_pred=Gradient.predict(X_train)
test_pred=Gradient.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  1.0
Accuracy on test data:  0.9730434782608696


In [72]:
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  1.0
F1 score on test data:  0.9730434782608696


In [73]:
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  1.0
Precision score on test data:  0.9730434782608696


In [74]:
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  1.0
Recall score on test data:  0.9730434782608696


In [75]:
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7365    0]
 [   0 3670]]
Confusion matrix on test data: 
 [[1816   19]
 [  43  422]]


How would the model have performed if the minority class in train data was not duplicated?

In [76]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X_prime,y,test_size=0.2,random_state=42)
print(X_train.shape)
print(X_test.shape)

(9200, 53)
(2300, 53)


In [77]:
from sklearn.ensemble import RandomForestClassifier
Forest=RandomForestClassifier(random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Forest

RandomForestClassifier(random_state=0)

In [78]:
from sklearn.metrics import accuracy_score
train_pred=Forest.predict(X_train)
test_pred=Forest.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  1.0
Accuracy on test data:  0.9721739130434782


In [79]:
from sklearn.ensemble import GradientBoostingClassifier
Gradient=GradientBoostingClassifier(random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Gradient

GradientBoostingClassifier(random_state=0)

In [80]:
train_pred=Gradient.predict(X_train)
test_pred=Gradient.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9767391304347826
Accuracy on test data:  0.9617391304347827


In [81]:
Gradient=GradientBoostingClassifier(n_estimators=210,learning_rate=0.15,max_depth=7,max_features="log2",
                                    random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
Gradient

GradientBoostingClassifier(learning_rate=0.15, max_depth=7, max_features='log2',
                           n_estimators=210, random_state=0)

In [82]:
train_pred=Gradient.predict(X_train)
test_pred=Gradient.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  1.0
Accuracy on test data:  0.9726086956521739


In [83]:
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  1.0
F1 score on test data:  0.9726086956521739


In [84]:
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  1.0
Precision score on test data:  0.9726086956521739


In [85]:
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  1.0
Recall score on test data:  0.9726086956521739


In [86]:
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7365    0]
 [   0 1835]]
Confusion matrix on test data: 
 [[1821   14]
 [  49  416]]


In [87]:
from sklearn import svm
SVM=svm.SVC(random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
SVM

SVC(random_state=0)

In [88]:
train_pred=SVM.predict(X_train)
test_pred=SVM.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9759782608695652
Accuracy on test data:  0.9721739130434782


In [111]:
SVM=svm.SVC(C=3,random_state=0).fit(X_train,np.asarray(y_train).reshape(y_train.shape[0],))
SVM

SVC(C=3, random_state=0)

In [112]:
train_pred=SVM.predict(X_train)
test_pred=SVM.predict(X_test)
print("Accuracy on train data: ",accuracy_score(y_train,train_pred))
print("Accuracy on test data: ",accuracy_score(y_test,test_pred))

Accuracy on train data:  0.9848913043478261
Accuracy on test data:  0.9765217391304348


In [113]:
print("F1 score on train data: ",f1_score(y_train,train_pred,average="micro"))
print("F1 score on test data: ",f1_score(y_test,test_pred,average="micro"))

F1 score on train data:  0.9848913043478261
F1 score on test data:  0.9765217391304348


In [114]:
print("Precision score on train data: ",precision_score(y_train,train_pred,average="micro"))
print("Precision score on test data: ",precision_score(y_test,test_pred,average="micro"))

Precision score on train data:  0.9848913043478261
Precision score on test data:  0.9765217391304348


In [115]:
print("Recall score on train data: ",recall_score(y_train,train_pred,average="micro"))
print("Recall score on test data: ",recall_score(y_test,test_pred,average="micro"))

Recall score on train data:  0.9848913043478261
Recall score on test data:  0.9765217391304348


In [116]:
print("Confusion matrix on train data: \n",confusion_matrix(y_train,train_pred))
print("Confusion matrix on test data: \n",confusion_matrix(y_test,test_pred))

Confusion matrix on train data: 
 [[7352   13]
 [ 126 1709]]
Confusion matrix on test data: 
 [[1824   11]
 [  43  422]]


## Summary

### Models from the "imblearn" library: 

|Model|Train data accuracy|Train data F1 score| Train data precision|Train data recall|
|-----|--------------------|------------------|---------------------|-----------------|
|BalancedRandomForestClassifier|0.999783|0.999782|0.999783|0.999783|
|BalancedBaggingClassifier with SVM|0.982717|0.982717|0.982717|0.982717|

|Model|Test data accuracy|Test data F1 score| Test data precision|Test data recall|
|-----|--------------------|------------------|---------------------|-----------------|
|BalancedRandomForestClassifier|0.971739|0.971739|0.971739|0.971739|0.971739|
|BalancedBaggingClassifier with SVM|0.973478|0.973478|0.973478|0.973478|0.973478|

### From the "sklearn" library with minority class in train data duplicated by 2 times:

#### 1. With default values for hyper-parameters:

|Model|Train data accuracy|Test data accuracy|
|-----|-------------------|------------------|
|RandomForestClassifier|1.0(100%)|0.97|
|GradientBoostingClassifier|0.971|0.96|
|Support Vector Machine|0.964567|0.972174|

#### 2. With tuned values for the hyper-parameters:

|Model|Train data accuracy|Train data F1 score| Train data precision| Train data recall|
|-----|-------------------|-------------------|---------------------|------------------|
|RandomForestClassifier|100%|100%|100%|100%|
|GradientBoostinClassifier|100%|100%|100%|100%|
|Support Vector Machine|0.987222|0.987222|0.987222|0.987222|

|Model|Test data accuracy|Test data F1 score|Test data precision|Test data recall|
|-----|------------------|------------------|-------------------|----------------|
|RandomForestClassifier|0.973913|0.973913|0.973913|0.973913|
|GradientBoostingClassifier|0.973043|0.973043|0.973043|0.973043|
|Support Vector Machine|0.979130|0.979130|0.979130|0.979130|

### From the "sklearn" library with no data resampling

#### 1. With default values for hyper-parameters:

|Model|Train data accuracy|Test data accuracy| 
|-----|-------------------|-------------------|
|RandomForestClassifier|100%|0.972173|
|GradientBoostingClassifier|0.976739|0.961739|
|Support Vector Machine|0.975978|0.972173|

#### 2. With tuned values for the hyper-parameters:

|Model|Train data accuracy|Train data F1 score|Train data precision|Train data recall|
|-----|-------------------|-------------------|--------------------|-----------------|
|GradientBoostingClassifier|100%|100%|100%|100%|
|Support Vector Machine|0.984891|0.984891|0.984891|0.984891|

|Model|Test data accuracy|Test data F1 score|Test data precision| Test data recall|
|-----|------------------|------------------|-------------------|-----------------|
|GradientBoostingClassifier|0.972608|0.972608|0.972608|0.972608|
|Support Vector Machine|0.976521|0.976521|0.976521|0.976521|

The best performance on test data is delivered by the Support Vector Machine model when the minority class in the train data is duplicated by 2 times to increase the model's exposure to the minority class. With a slight tuning of just two of the hyper-parameters, the SVM model deliver a 97.913% accuracy on test data. The second best performance on test data is delivered by the SVM model as well, with no data resampling. In this case, by fine-tuning just 1 hyper-parameter, the SVM model delivered a 97.652% accuracy on test data. So, duplicating the minority class in train data by 2 times increased the best model's performance further by 0.261%. That number by itself does not sound like a lot. However, it we put it in the context of a dataset with more than 10,000 rows of data, an increase of 0.261% in test data accuracy means that the model is able to make **2,610** more correct predictions. 

Besides the quick data resampling technique that I showed above, feature selection can also be an effective strategy to tackle class imbalance and improve model performance. The entire dataset has 178 potential predictor variables, I used only 53 of them. That is less than 30%. I implemented the SelectKBest with the "f_classif" function and set the threshold for feature score at 8. I found that lowring the feature score threshold further did not contribute to any improvement in model performance. On the contrary, including more less relevant features compromised even the best model's performance.